docs/proposals/21-global-failover.md

```diff
@@ -0,0 +1,169 @@
+authors:
+- "@jia-jerry"
+- "@EmoinLanyu"
+reviewers:
+- "@main-reviewer-1"
+- "@main-reviewer-2"
+---
+
+# GEP-21: Fail Over Multi-AZ Services by Isolating Unhealthy AZ
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+    - [New Changes](#new-changes)
+    - [Control Failover / Failback Process](#control-failover--failback-process)
+    - [Failover Workflow](#failover-workflow)
+    - [Failback Workflow](#failback-workflow)
+- [Future Plan](#future-plan)
+
+## Summary
+
+This `GEP` introduces an operating procedure allowing for Gardener operators to isolate an unhealthy AZ of a `Shoot` cluster, and consequently fail over services running inside to healthy AZs. This procedure consists of two steps: removing `MachineDeployment` in the unhealthy AZ, and applying `SubnetIsolation` to the unhealthy AZ. The `SubnetIsolation` is a new extension resource that is reconciled by the respective provider extension controller.
+
+Moreover, it covers a failback procedure for services with multi-AZ setup once the unhealthy AZ is recovered.
+
+## Motivation
+
+Several outages in past years have shown that AZ-specific degradations / failures in hyperscaler services result in significant issues for services. There are multiple patterns of known issues.
+
+1. Some services operate multiple instances of each components redundantly within a single AZ. This will likely provide uptimes approaching 99.5%, but such services are subject to AZ failures and remain offline until the underlying AZ is restored.
+2. Often, multi-AZ HA services are able to stay available during AZ-specific degradations, such as slow disk IO, network lantency, etc. But their api performance may be downgraded tenfold.
+3. In some edge cases, multi-AZ HA services may run into inconsistent state due to unexpected AZ degradations / failures. Manual operations are needed in order to recover the problematic service back to normal.
+
+A global failover, i.e. not using an unhealthy AZ any longer, will bring benefits and reduce customer disruptions. For example, in case of pattern 2, by isolating the unhealthy AZ, service will be forced to failover, all traffic will be routed to healthy AZs, and consequently service performance will be back to normal quickly. In case of pattern 3, manual operations are inevitable for recovering services in inconsistent state, however forced failovers of other services can eliminate new inconsistent state during a long period of hyperscaler failures.
+
+However, we shouldn't fail over a `Shoot` cluster with single AZ HA services.
+
+### Goals
+
+Reduce customer disruptions during a long period of hyperscaler failure by isolating unhealthy AZ, enforcing a full zonal outage and triggering failovers / failbacks of services in a controlled way.
+
+### Non-Goals
+
+- Help services to achieve multi-AZ HA.
+- Rebalance `Pod`s during / after failback.
```

Comment:
OK, but if you go into de-balancing it, you also should go into re-balancing.

BTW, exactly that was the reason why we never considered this approach much:
- Find out that a zone is "broken"
- Take the automated (we never considered manual and I think that's a big no-no) decision to re-balance (beyond what Kubernetes provides out-of-the-box)
- And back again, i.e. find out that a zone is "healthy" again
- Reverse everything and re-balance to survive the next zone outage, possible of another zone

Reply:
Regarding the de-balancing topic, let's discuss three scenarios.
1. For pods with PV attached, no de-balancing happens.
2. For pods without PV over multi-AZ using node anti-affinity patterns, no de-balancing happens.
3. For other pods without PV, they will be scheduled to healthy zones, which leads to a de-balancing situation. If we do nothing after AZ is restored, the pods can eventually become balanced in case of pods rescheduling or updates. It may take a long time, e.g. a couple of weeks. In future, we may consider a solution to do re-balancing faster.

Reply:
So, the proposal is mostly useless for services with persistence anyhow?

And for the others, why do you need a red button? The nodes will become NotReady, flexible workload will anyhow be scheduled to other zones with healthy nodes automatically.